src/DiffBuilder/Diff.php

```diff
@@ -0,0 +1,116 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Phpcq\RepositoryBuilder\DiffBuilder;
+
+use LogicException;
+use Phpcq\RepositoryBuilder\Repository\Tool;
+
+final class Diff implements DiffInterface
+{
+    /**
+     * @var DiffInterface[]|null
+     */
+    private ?array $differences;
+
+    /**
+     * @param Tool[]|null $old Tools indexed by tool name.
+     * @param Tool[]|null $new Tools indexed by tool name.
+     *
+     * @return Diff|null
+     */
+    public static function diff(?array $old, ?array $new): ?Diff
+    {
+        if (null === $old && null === $new) {
+            throw new LogicException('new value and old value must not both be null.');
+        }
+
+        // New repository, add all tools as new.
+        if (null === $old) {
+            $changes = [];
+            foreach ($new as $tool) {
+                $changes[$tool->getName()] = ToolAddedDiff::diff($tool);
+            }
+            return new Diff($changes);
+        }
+
+        // Tool got removed, add all versions as removed.
+        if (null === $new) {
+            $changes = [];
+            foreach ($old as $tool) {
+                $changes[$tool->getName()] = ToolRemovedDiff::diff($tool);
+            }
+            return new Diff($changes);
+        }
+
+        return self::deepCompare($old, $new);
+    }
```

Comment:
I prefer 3 different constructor methods here:

```php
public static function removed(array $old) : ?Deep
public static function created(array $new) : ?Deep
public static function diff(array $old, array $new) : ?Deep
```

We already have a diff builder which can select the right constructor here.

Reply:
Changed in 041a96a0d6666d246c4f7ec2bad138d59ffd82c4